pyro/distributions/spanning_tree.py

```diff
@@ -0,0 +1,418 @@
+from __future__ import absolute_import, division, print_function
+
+import itertools
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import lazy_property
+
+from pyro.distributions.torch_distribution import TorchDistribution
+
+
+class SpanningTree(TorchDistribution):
+    """
+    Distribution over spanning trees on a fixed number ``V`` of vertices.
+
+    :meth:`log_prob` is implemented using Kirchoff's theorem and a
+    Cholesky-based computation of log determinant.
+
+    :meth:`sample` is implemented using MCMC run for a small number of steps
+    after being initialized by a cheap approximate sampler. This sampler is
+    approximate and cubic time. This is much faster than the classic
+    Aldous-Broder sampler [1,2] for graphs with large mixing time. Recent
+    research [3,4] proposes samplers that run in sub-matrix-multiply time but
+    are more complex to implement.
+
+    :meth:`enumerate_support` is implemented for trees with up to 6 vertices.
+
+    **References**
+
+    [1] `Generating random spanning trees`
+        Andrei Broder (1989)
+    [2] `The Random Walk Construction of Uniform Spanning Trees and Uniform Labelled Trees`,
+        David J. Aldous (1990)
+    [3] `Sampling Random Spanning Trees Faster than Matrix Multiplication`,
+        David Durfee, Rasmus Kyng, John Peebles, Anup B. Rao, Sushant Sachdeva
+        (2017) https://arxiv.org/abs/1611.07451
+    [4] `An almost-linear time algorithm for uniform random spanning tree generation`,
+        Aaron Schild (2017) https://arxiv.org/abs/1711.06455
+
+    :param torch.Tensor edge_logits: A tensor of length ``V*(V-1)//2``
+        containing logits (aka negative energies) of all edges in the complete
+        graph.  Edges in the complete graph are ordered:
+        ``(0,1), (0,2), (1,2), (0,3), (1,3), (2,3), (0,4), (1,4), (1,5), ...``
+    :param dict sampler_options: An optional dict of sampler options including:
+        ``mcmc_steps`` defaulting to a single MCMC step (which is pretty good);
+        ``initial_edges`` defaulting to a cheap approximate sample;
+        ``backend`` one of "python" or "cpp", defaulting to "python".
+    """
+    arg_constraints = {'edge_logits': constraints.real}
+    support = constraints.positive_integer
+    has_enumerate_support = True
+
+    def __init__(self, edge_logits, sampler_options=None, validate_args=None):
+        if edge_logits.is_cuda:
+            raise NotImplementedError("SpanningTree does not support cuda tensors")
+        K = len(edge_logits)
+        V = int(round(0.5 + (0.25 + 2 * K)**0.5))
+        assert K == V * (V - 1) // 2
+        E = V - 1
+        event_shape = (E, 2)
+        batch_shape = ()
+        self.edge_logits = edge_logits
+        super(SpanningTree, self).__init__(batch_shape, event_shape, validate_args=validate_args)
+        if self._validate_args:
```

Comment:
nit: consider moving this check earlier (after computing K).

Reply:
The reason I put it here is because `self._validate_args` is only defined after `super(...).__init__()` (recall that `self._validate_args` is true if either `validate_args is True` or `validate_args is None and Distribution._validate_args is True`).

Reply:
Ahh..thanks for clarifying!